### PR TITLE
Add .netrc support

### PR DESCRIPTION
**Pull request by mkosiba**
This is to support something like an internal mirror (such as Artifactory).
It's possible to use an one by hand-crafting a `repo.yaml` file, but things get tricky if such a mirror requires authentication.
This adds support for basic auth driven by a .netrc file.

### CODE REVIEW

File: pkg/repo/fetch.go [hunk 1]
```diff
@@ -229,6 +232,41 @@ func fileGet(filename string) (*http.Response, error) {
...
+
+func getNetrc() (*netrc.Netrc, error) {
+	if netrcCached != nil {
+		return netrcCached, nil
+	}
+	usr, err := user.Current()
+	if err != nil {
+		return nil, fmt.Errorf("getting current user: %w", err)
+	}
+	n, err := netrc.Parse(filepath.Join(usr.HomeDir, ".netrc"))
```
[the remark this answers is not in the document]
Comment (mkosiba) [reply pkg/repo/fetch.go:245]: AFAICT this part of bazeldnf runs in the repository rule / loading phase, which is "allowed" to poke at local files and perform network calls. 
Also, this is reading the .netrc file to supply credentials to a http GET request. The GET request itself is also not hermetic.

[the remark this answers is not in the document]
Comment (kellyma2) [reply pkg/repo/fetch.go:245]: Moreso what I was getting at is that this dependency is opaque to everything that consumes bazeldnf itself.  ie: how would you identify this as something that you build depends on?

[the remark this answers is not in the document]
Comment (mkosiba) [reply pkg/repo/fetch.go:245]: Bazel already defaults to reading ~/.netrc for the built in http_archive rules (see https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/http.bzl#L144 and https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/utils.bzl#L460 which show how it's implemented) without any explicit parameters from the user. This is well established behaviour.

Actually, I had another look through the code and I'll go back on my previous statement. I don't think this code path runs as part of the build at all:
* the `bazeldnf` target is only exposed as a user convenience (though, yea, sure, someone could run it as part of their custom rule or a genrule, but that's on them)
* the two rules that invoke `bazeldnf` are `rpm2tar` and `tar2files`, neither of which use the `repo` package, so they shouldn't hit this code path,
* the `rpm` rules use Bazel's downloader to download the files.

In other words these changes only affect `bazeldnf` invocations that are performed by the user locally prior to the build (like `bazeldnf fetch`) and so build hermeticity concerns shouldn't come into play at all.

[the remark this answers is not in the document]
Comment (manuelnaranjo) [reply pkg/repo/fetch.go:245]: This will only be used while resolving dependencies, all the rest of the time Bazel will use the rpm repo rule which under the hood uses the same mechanisms as http_archive

File: go.mod [hunk 2]
```diff
@@ -17,6 +17,7 @@ require (
 	github.com/golang/protobuf v1.5.2 // indirect
 	github.com/google/go-cmp v0.5.9 // indirect
 	github.com/inconshreveable/mousetrap v1.0.1 // indirect
+	github.com/jdx/go-netrc v1.0.0 // indirect
```
Comment (manuelnaranjo) [inline go.mod:20]: Mind to rebase the branch and add some documentation so we can merge?

Could the path be configurable through some env variable? And then some test that does some url rewriting or something that requires no human interaction be created?

File: pkg/repo/fetch.go [hunk 3]
```diff
@@ -234,6 +237,41 @@ func fileGet(filename string) (*http.Response, error) {
...
+	if err == nil {
+		netrcCached = n
+	}
+	return n, err
```
Comment (manuelnaranjo) [inline pkg/repo/fetch.go:254]: I don't think you should return the err in this case, or maybe you should check for the file existing first?

Comment (mkosiba) [reply pkg/repo/fetch.go:260]: I'd expect there to be a parse error if the file is malformed. Are you suggesting we ignore that instead?

Comment (mkosiba) [reply pkg/repo/fetch.go:260]: oh, nevermind, I looked at the implementation and the only error it returns is if the file doesn't exist

Comment (manuelnaranjo) [reply pkg/repo/fetch.go:260]: If the file is not there, which is most likely the case, shouldn't make the tool fail

File: pkg/repo/fetch.go [hunk 4]
```diff
@@ -234,6 +238,85 @@ func fileGet(filename string) (*http.Response, error) {
...
+			return nil, fmt.Errorf("getting current user: %w", err)
+		}
+		netrcPath = filepath.Join(usr.HomeDir, ".netrc")
+	}
+
+	if netrcPath != "" {
+		n, err := netrcCache.Read(netrcPath)
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				log.Warnf("Error reading netrc from %s: %v", netrcPath, err)
```
[the remark this answers is not in the document]
Comment (mkosiba) [reply pkg/repo/fetch.go:282]: > If the user doesn't set NETRC and doesn't have a .netrc then this warning will always trigger miss leading users

That's what the `!errors.Is(err, os.ErrNotExist)` is for. This should only trigger if the file exists but, say, doesn't have read permissions.
